drivers/ws281x/timer_gpio_ll.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2023 Christian Amsüss
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ws281x
+ *
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the WS281x abstraction based on GPIO_LL and timers
+ *
+ * When this is used, the whole @ref ws281x_write_buffer operation is conducted
+ * in a single interrupts-disabled run. As it takes about 1us per bit (eg.
+ * 0.3ms on a 100-LED strip), the use of this module needs to be carefully
+ * evaluated against other real-time requirements.
+ *
+ * (Letting the trailing pause run with interrupts enabled, or even enabling
+ * interrupts between bits, is an option that is being considered for future
+ * expansion).
+ *
+ * @author      Christian Amsüss <chrysn@fsfe.org>
+ *
+ * @}
+ */
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "irq.h"
+#include "time_units.h"
+#include "periph/timer.h"
+#include "periph/gpio_ll.h"
+
+#include "ws281x.h"
+#include "ws281x_params.h"
+#include "ws281x_constants.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/* (+ NS_PER_SEC - 1): Rounding up, as T1H is the time that needs to distinctly
+ * longer than T0H.
+ *
+ * Then adding +1 extra, because the spin loop adds another layer of jitter. A
+ * more correct version would be to add the spin loop time before rounding (and
+ * then rounding up), but as that time is not available, spending one more
+ * cycle is the next best thing to do. */
+const int ticks_one = ((uint64_t)WS281X_T_DATA_ONE_NS * (uint64_t)WS281X_TIMER_FREQ + NS_PER_SEC - 1) / NS_PER_SEC + 1;
```

Comment:
```suggestion
const int ticks_one = ((uint64_t)WS281X_T_DATA_ONE_NS * WS281X_TIMER_FREQ + NS_PER_SEC - 1)
                    / NS_PER_SEC + 1;
```

one cast should be enough